File: output/sonarqube_analysis/copilot_solutions/python/results/q37-longest-valid-parentheses.py
# https://leetcode.com/problems/longest-valid-parentheses/
# Problem: Longest Valid Parentheses

#Problem Description:
#Given a string containing just the characters '(' and ')', return the length of the longest valid (well-formed) parentheses substring.

# Constraints:
# 0 <= s.length <= 3 * 10^4
# s[i] is '(', or ')'.

#Code Structure:
class Solution(object):
    def longestValidParentheses(self, s):
        """
        :type s: str
        :rtype: int
        """
        stack = []
        max_length = 0
        last_invalid_index = -1
        
        for i, char in enumerate(s):
            if char == '(':
                stack.append(i)
            else:
                if stack:
                    stack.pop()
                    if stack:
                        max_length = max(max_length, i - stack[-1])
                    else:
                        max_length = max(max_length, i - last_invalid_index)
                else:
                    last_invalid_index = i
        
        return max_length                             
    

File: output/sonarqube_analysis/copilot_solutions/php/results/q37-longest-valid-parentheses.php
<?php
/** https://leetcode.com/problems/longest-valid-parentheses */
// Problem: Longest Valid Parentheses

//Problem Description:
//Given a string containing just the characters '(' and ')', return the length of the longest valid (well-formed) parentheses substring.

// Constraints:
// 0 <= s.length <= 3 * 10^4
// s[i] is '(', or ')'.


//Code Structure:
class Solution {

    /**
     * @param String $s
     * @return Integer
     */
    function longestValidParentheses($s) {
        $maxLength = 0;
        $stack = new SplStack();
        $stack->push(-1); // Base index for valid substring

        for ($i = 0; $i < strlen($s); $i++) {
            if ($s[$i] === '(') {
                $stack->push($i);
            } else {
                $stack->pop(); // Pop the last open parenthesis index
                if ($stack->isEmpty()) {
                    $stack->push($i); // Push current index as base
                } else {
                    $maxLength = max($maxLength, $i - $stack->top());
                }
            }
        }

        return $maxLength;
    }
}
    